### PR TITLE
Checkpoint Training Model

### PR DESCRIPTION
**Pull request by JustinBakerMath**
Basic checkpointing for training based on the synchronized `validation_loss` performance metric.

Can be turned on by adding
```
"Training" : {
    "Checkpoint" : true,
}
```
with additional `checkpoint_freq` argument for the frequency of checkpointing (`default=10`)
```
"Training" : {
    "Checkpoint" : true,
    "checkpoint_freq" : 10,
}
```

### CODE REVIEW

File: tests/inputs/ci.json [hunk 1]
```diff
@@ -68,6 +68,8 @@
             "perc_train": 0.7,
 	    "EarlyStopping": true,
 	    "patience": 10,
+        "Checkpoint": true,
```
Comment (allaffa) [inline tests/inputs/ci.json:71]: In terms of indentation, it is best that checkpoint be encapsulated within "Training"

File: tests/inputs/ci.json [hunk 2]
```diff
@@ -66,8 +66,10 @@
         "Training": {
             "num_epoch": 100,
             "perc_train": 0.7,
-	    "EarlyStopping": true,
-	    "patience": 10,
+            "EarlyStopping": true,
```
Comment (allaffa) [inline tests/inputs/ci.json:69]: Would you mind adding this section to all the CI.json file where at least 10 epochs are performed?

**Review by allaffa — CHANGES_REQUESTED** [review]
We should add the additional input to all the JSON file of CI tests that perform at least 10 epochs.

**Comment by JustinBakerMath** [comment]
After revisiting this: `checkpoint_freq` and `min_perf_metric` aren't ideally compatible.

If the goal is to use `min_perf_metric` to checkpoint, then there should be a *warmup* rather than a frequency.

I've refactored this, replacing `checkpoint_freq` with `checkpoint_warmup` for a *warmup* to checkpointing.

Two print lines are also included
- (every epoch) indicate best performance metric 
- (upon saving) indicate new checkpoint with performance metric.

File: hydragnn/utils/model.py [hunk 3]
```diff
@@ -159,3 +159,39 @@ def __call__(self, val_loss):
...
+        self.count = 1
+        self.warmup = warmup
+        self.path = path
+        self.name = name
+        self.min_perf_metric = float("inf")
+        self.min_delta = 0
+
+    def __call__(self, model, optimizer, perf_metric):
+
+        if (perf_metric > self.min_perf_metric + self.min_delta) or (
```
[the remark this answers is not in the document]
Comment (allaffa) [reply hydragnn/utils/model.py:189]: The call to `save_model` has no input parameter that varies over the iteration count. How do you make sure that different checkpoints are saved in directories with different path names?

**Comment by allaffa** [comment]
@jychoi-hpc 

Hi Jong. Justin implemented a checkpoint restart which saves a copy of the model when the validation loss function gets reduced at a new iteration. A warm-up waiting is applied at earlier iterations of the training to avoid saving too many models at the beginning. 
I was wondering if the PR may cause too many bottlenecks with significant slow-down for training. 
Any thoughts on this?

**Review by jychoi-hpc — APPROVED** [review]
I checked on Summit. It works for me. Since users can control (delay, etc) through json file, I think it is ok.
